src/query/processor/executor/IntraPartitionParallelExecutor.h

```diff
@@ -0,0 +1,284 @@
+#define INTRA_PARTITION_PARALLEL_EXECUTOR_H
+
+#include <vector>
+#include <memory>
+#include <mutex>
+#include <condition_variable>
+#include <atomic>
+#include <queue>
+#include <functional>
+#include <thread>
+#include <future>
+#include <algorithm>
+#include <unordered_set>
+#include "../../../performance/metrics/StatisticsCollector.h"
+
+/**
+ * Thread-safe node data structure for parallel processing
+ * Holds pre-loaded node data to avoid thread safety issues with NodeManager
+ */
+struct PreloadedNodeData {
+    std::string nodeId;
+    std::string partitionId;
+    std::map<std::string, std::string, std::less<>> properties;
+
+    PreloadedNodeData(const std::string& id, const std::string& partId,
+                      const std::map<std::string, std::string, std::less<>>& props)
+        : nodeId(id), partitionId(partId), properties(props) {}
+};
+
+/**
+ * Thread-safe buffer for managing parallel access to partition data
+ */
+template<typename T>
+class ThreadSafeBuffer {
+ private:
+    std::queue<T> buffer;
+    mutable std::mutex bufferMutex;
+    std::condition_variable cv;
+    bool finished = false;
+
+ public:
+    void push(const T& item) {
+        std::scoped_lock lock(bufferMutex);
+        buffer.push(item);
+        cv.notify_one();
+    }
+
+    bool pop(T& item, std::chrono::milliseconds timeout = std::chrono::milliseconds(100)) {
+        std::unique_lock lock(bufferMutex);
+        return cv.wait_for(lock, timeout, [this, &item] {
+            if (!buffer.empty()) {
+                item = buffer.front();
+                buffer.pop();
+                return true;
+            }
+            return finished;
+        });
+    }
+
+    void setFinished() {
+        std::scoped_lock lock(bufferMutex);
+        finished = true;
+        cv.notify_all();
+    }
+
+    size_t size() const {
+        std::scoped_lock lock(bufferMutex);
+        return buffer.size();
+    }
+
+    bool empty() const {
+        std::scoped_lock lock(bufferMutex);
+        return buffer.empty() && finished;
+    }
+};
+
+/**
+ * Represents a chunk of work for parallel processing
+ */
+struct WorkChunk {
+    long startIndex;
+    long endIndex;
+    size_t estimatedSize;
+
+    WorkChunk(long start, long end, size_t size)
+        : startIndex(start), endIndex(end), estimatedSize(size) {}
+};
+
+/**
+ * Dynamic thread pool that adapts to system capabilities
+ * Automatically detects CPU core count and creates optimal number of workers
+ */
+class DynamicThreadPool {
+ private:
+    std::vector<std::thread> workers;
+    std::queue<std::function<void()>> tasks;
+    std::mutex queueMutex;
+    std::condition_variable condition;
+    std::atomic<bool> stop{false};
+    int optimalWorkerCount;
+
+ public:
+    DynamicThreadPool();
+    ~DynamicThreadPool();
+
+    template<class F, class... Args>
+    auto enqueue(F&& f, Args&&... args)
+        -> std::future<typename std::result_of<F(Args...)>::type>;
```

Comment:
Replace `std::result_of` with `std::invoke_result_t` (portable to C++17/20) in `DynamicThreadPool::enqueue`